labrad/decorators.py

```diff
@@ -21,202 +21,140 @@
 
 from __future__ import absolute_import
 
-from functools import wraps
-from inspect import getargspec
+import functools
+import inspect
+import itertools
 
-from twisted.internet.defer import inlineCallbacks
+import twisted.internet.defer as defer
 
 from labrad import types as T, util
 
-def _isGenerator(f):
-    """Check to see whether f is a generator.
-
-    See the documentation on code objects at:
-    http://docs.python.org/ref/types.html
-    """
-    return bool(f.func_code.co_flags & 0x20)
-
 def _product(lists):
     """Return the cartesian product of a list of lists."""
     if not len(lists): return [[]]
     return [[h] + t for h in lists[0] for t in _product(lists[1:])]
 
-class Setting(object):
-    def __init__(self, func):
-        self.func = func
+def setting(lr_ID, lr_name=None, returns=[], unflatten=True, **params):
+    def decorator(func):
+        return Setting(func, lr_ID, lr_name, returns, unflatten, **params)
+    return decorator
 
-    def getRegistrationInfo(self):
-        return (long(self.ID), self.name, self.description,
-                self.accepts, self.returns, self.notes)
+class Setting(object):
+    """
+    Represent a single setting within a Server object.
 
-    def handleRequest(self, c, data):
-        return self.func(c, data)
+    args:
 
-def setting(lr_ID, lr_name=None, returns=[], lr_num_params=2, **params):
-    """Mark a server method as a remotely-accessible setting.
+    func:       Function to implement the setting.
+    lr_ID:      The ID number of the setting used in the labrad protocol
+    lr_name:    The setting name.  By default, this is derived from the function name
+    returns:    The possible labrad type strings returned by the setting
 
-    The only required parameter is an integer ID.  You may
-    also provide a name to override the name of the decorated
-    function.  In addition, accepted types for each of the setting
-    parameters may be provided as named parameters with a list of
-    strings of allowed types.
+    **kw:       Additional keyword arguments indicate arguments to the setting.  Each keyword
+                should be associated with one or more labrad type strings indicating data types
+                it accepts.
```

Comment:
"associated with" is a little vague, and it's also not clear that the names of these arguments should match the names of arguments to the server method being decorated. I would suggest rewording to make this clear.